torchtune/generation/_generation.py

```diff
@@ -325,6 +321,11 @@ def generate(
 
     # keeps track at a high level if we've already hit a stop token in a sequence so we can early stop
     stop_token_reached = torch.zeros(bsz, dtype=torch.bool, device=prompt.device)
+    stop_tokens = (
+        torch.tensor(stop_tokens, device=prompt.device, dtype=tokens.dtype)
```

Comment:
Not sure why but outputs from the model on my device were `int32`. The prompt is `int64` and `torch.isin` was complaining when `stop_tokens.dtype != tokens.dtype` later on. I wasn't ready for further investigation.

Reply:
carry on

Reply:
```python
import torch
for dvc in ["cpu", "cuda"]:
    print(dvc)
    device = torch.device("cuda")
    x = torch.arange(10, device = device, dtype=torch.int64)
    y = torch.arange(10, device = device, dtype=torch.int32)
    torch.isin(x, y)
```
So fine on CPU/CUDA
BUT this seems like expected behaviour on MPS? There's actually a test for this in core
https://github.com/pytorch/pytorch/blob/e5a57932f06746a7f6779a97b4e11394b614fa81/test/test_mps.py#L8512

Reply:
yeah there's a check for it in the MPS implementation https://github.com/pytorch/pytorch/blob/e5a57932f06746a7f6779a97b4e11394b614fa81/aten/src/ATen/native/mps/operations/TensorCompare.mm#L297

Reply:
Maybe because the underlying MPS call doesn't support different datatypes?
https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/equal(_:_:name:)?language=objc

Reply:
This is probably it and there's NO WAY I'm loading up XCode to try it out. I hope you are contented.
https://xkcd.com/356/